### DISCUSSION TITLE
Going serverless with SQL?

### DISCUSSION THREAD

**Discussion by michael** [vercel/community#96] [Help] [answered]
I’m realising the potential of going Serverless, since my web platform is pretty simple, and I love front end and backend dev but I REALLY hate managing servers and spend time in admin consoles - it eats a lot of my time.

While the solution is pretty clear for Images+Videos (S3, ImgIX, Mux.com), I’m puzzled about how to solve the SQL part of the equation.

In my case, I guess what I want is a Postgres DB exposed as a scalable HTTP service, so i can throw queries (+ transactions) at it from my Next.js serverless functions deployed on Vercel.

I tried Amazon Aurora Serverless that offers that, but there there is no way to connect to the DB directly from public (I need that for other reasons).

I’m also wondering how I can ensure that there’s minimum response time, when I’m hitting the DB. So I’m worried even if I found a service that does that for me, the performance wouldn’t be good as it is not close to Vercel's infrastructure.

What would you recommend me to do?

If Vercel could promote a preferred way to deploy an SQL backed-app serverless, that would be fantastic. Ideal would be a Postgres Option on Vercel directly, so I can just stop thinking about it. That would be such a killer feature! :)

Thanks for the help,
Michael

**Comment by mcsdevv** [comment]
Hi Michael,

The most popular options I've used and encounter others using are Postgres on Digital Ocean, this is almost always because they have built-in connection pooling you can enable with a click using `pgbouncer`. For a serverless app, this is very important so you do not exhaust the connections available.

Another option is Supabase, they handle the setup and allow access via PostgREST which also helps to remove (or considerably reduce) the concern about exhausting connections.

You can find further information here: https://vercel.com/docs/solutions/databases

**Comment by michael** [comment]
Thanks @mcsdevv! I was aware of the connection pooling feature of DO. Currently I'm using Digitalocean Apps and a Postgres DB hosted there as well. That's a sound classic server-based deployment I think. Also the DB is close to the Server, so DB queries come back fast. However, it doesn't have the serverless properties of scaling up or shutting down on low load.

Or is what you are trying to say "use a Digitalocean database but connect to it remotely from Vercel"? However then I'd be worried about the delay as the data needs to travel quite far.

PostgREST is an interesting take. I haven't heard of it before. I'm a bit hesitant as it adds a DSL, rather than just allowing to perform queries. Want to have as little layers as possible. Maybe there's something similar that "just" exposes the DB via http? Anyway's I'd like to avoid managing a separate microservice as also that needs to be deployed and makes my app more complex. 

So I think for Vercel + SQL to work really smoothly for the developer (and be fast) it somehow would have be integrated into Vercel. Like tick a box somewhere and Vercel provisions a Database cluster on their infrastructure and makes it available to Next.js serverless functions via an HTTP API. Just dreaming... 🙈

Glad to hear a service such as Supabase exists. I also had a look at Firebase but it's too opinionated for my use-cases, so i rather hand-craft my simple email token based authentication.

**Reply by kiwicopple** [reply]
Supabase cofounder here. 

> PostgREST is an interesting take. I haven't heard of it before. I'm a bit hesitant as it adds a DSL

PostgREST also gives you the ability to query views and stored procedures. So if you use Supabase you can [select](https://supabase.io/docs/client/select) from a view or an [rpc](https://supabase.io/docs/client/rpc). As @mcsdevv mentioned above, this will remove any problems with connection pooling as PostgREST consumes a single connection to your DB.

> Maybe there's something similar that "just" exposes the DB via http?

On the roadmap :)

**Comment by michael** [comment]
@mcsdevv I'd give Supabase a spin, as I'm curious. :) So if I deploy that example project on Vercel that means it talks with Supabase's HTTP endpoints. But that would then exactly have that performance penalty I mentioned right? So for this to be really fast in DB access either Supabase needs to offer a deployment option on their own, or Vercel+Supabase have an agreement to use the same data center? :D

**Reply by leerob** [reply]
You can [specify which region you want to deploy your serverless functions to](https://vercel.com/docs/configuration#project/regions) on paid Vercel plans. Then you can place your functions near your database. For example, both in us-east-1.

**Reply by michael** [reply]
I see. I this a common way to do it? I'm not too experienced in this, but does it perform sufficiently well, e.g. when compared to my serverful (what a word!) Digitalocean Server+DB solution. I can never be sure where the servers are exactly located. Of course it would be beneficial to know that (e.g. Supabase) runs on AWS, like Vercel does and then choose the same region.

**Reply by leerob** [reply]
Yes, it's common (and recommended!) to co-locate your server and your functions. You want to reduce that latency as much as possible!

**Reply by kiwicopple** [reply]
Hey @michael, Supabase cofounder here. We're on AWS, so the regions should match up with Vercel's. We will also be working on replication at some point this year, so you will be able to spread your data around the world to reduce latency

**Comment by michael** [comment, marked as answer]
Okay to summarize - I think these are my options:

1.  **Regular postgres** connections but with an Endpoint that has Connection pooling (pg-bouncer) as offered by DigitalOcean
1. **AWS Aurora Serverless** and using their Data API for Aurora Serverless
1. **Supabase** - really great to see there's a Firebase-ish offering backed by Postgres (just need to see if it's a good fit for my particular use-case)
1. **PostgREST** roll on my own (needs a separate deployment)
1. **Roll my own small http API** to proxy the DB (more lightweight than PostgREST but also needs a separate deployment)

Notes:
- As @leerob states, it is important to co-locate DB and functions. It doesn't need to be the same provider though.

I have a tendency towards 1., since that's the environment I'm most familiar with and have everything more or less running already, just need to port to Next.js. And possibly I make a little query interface abstraction, so I'm prepared to switch out pg-promise with something like AWS Aurora Data API.

Many thanks to everyone so far! If you can think of more options please let me know. I'm really curious, so I can make an informed decision in the coming days.

@kiwicopple I'll leave some more questions in the Supabase related thread.

**Comment by arcataroger** [comment]
Depending on what your app is and what needs to go into the database, it may be possible to rearchitect your site so you don't have to worry about scaling a DB as often? Do any of the scenarios apply to you?

**(Best case... write once, read often)**
Your staff, not users, updates the database: cache DB output at build time, and users' browsers only ever hit the cached HTML/JSON output, never the DB itself

**(Write rarely, read often)**
Most of your visitors are read-only, but a few need to write to the DB (such as browsers vs purchasers on an ecommerce site): Since most of your visitors will be browsers that don't need to update the DB, just see its query outputs, you can cache the query outputs behind a highly-scalable key-value store like [Redis](https://www.upstash.com/) or [Cloudflare Workers KV](https://www.cloudflare.com/products/workers-kv/). Scaling those up is much simpler than trying to scale a SQL database.

Alternatively (and probably simpler), you could cache the DB lookups behind a [cached serverless function](https://vercel.com/docs/serverless-functions/edge-caching) using a REST API. The function does the lookup from the DB and returns a JSON result set, but has its own built-in cache. If the next identical request comes in within X minutes, the function would just serve the response straight from its own cache instead of the DB. That way, the serverless function can scale from 10 to 10,000 requests in a second without increasing the load on your origin database.

With either the KV store or a caching function solution, though, you'll have to keep in mind the time it takes to propagate database updates (in terms of how quickly a cluster can replicate changes, or serverless propagation times and TTLs).

In either case, transactions would still hit the DB, but read-only queries are cached and don't need to hit the DB often (whatever you set your TTL to). And for those read-only users, you'd effectively negate the distance between Vercel and the DB, since most of them would only see the cached output straight from the serverless function (which, in Vercel's case, is hosted on Amazon Lambda). You'll have to [pay extra for multi-region serverless function hosting](https://vercel.com/docs/configuration#project/regions).

Lastly, with any of the above, you can also use [incremental static regeneration](https://nextjs.org/blog/next-9-5#stable-incremental-static-regeneration) to help build/"bake" updates into the static site output every few minutes, so users get even faster performance. It's another layer of caching on top of the serverless/KV caching.

**(Much harder: Write often, read often)** If you need real-time transactions for all your visitors (something like a social  network), that's much harder, and I'll leave that for the experts at Supabase or Firebase :) Maybe you've already identified that this is what you need, and if so, I apologize!

**Comment by yehudamakarov** [comment]
I'm always trying to deploy my database to be as low latency as possible. Sometimes this includes using a cloud provider where my backends are and using a virtual network amongst other things. But I'm not always positive as to which elements decrease latency the most. And if the regions being close always result in less latency etc. 

I would love if vercel had a way to deploy database for you that was wired up properly with the functions for your deployment. But if I deploy to digital ocean in a close region, how much extra latency are we talking?

And if I want to "solve" this by deploying a next app on my own, I don't know if I'll necessarily do "edge functions" in which case does it make sense to focus on the db latency at the loss of the edge compute? I don't think so. Hm. Thinking out loud.

**Reply by leerob** [reply]
Vercel does have integrations that allow you to deploy both your frontend and connect to your database:

- https://vercel.com/integrations/planetscale
- https://vercel.com/integrations/supabase

**Reply by yehudamakarov** [reply]
ok cool, let me check stuff out there. thanks
